### ISSUE TITLE
fit with data iterator

### ISSUE THREAD

**Issue by deadsoul44** [dmlc/xgboost#7696]
Hello,

I tried to train a model with a data generator like in the following example:

https://github.com/dmlc/xgboost/blob/6762c45494f5eabdf3c40912cf5c005ced6d332f/demo/guide-python/quantile_data_iterator.py

But, it failed while building `DeviceQuantileDMatrix` from iterator due to OOM. 

I know that model training with a data generator is not easy due to binning. I wonder if there can be a workaround. For example, we can provide upper and lower bounds for each feature in iterator. In each leaf, bins can be pre-defined with this information and the split information from upper nodes. So that at each batch iteration, calculated loss values can be put in the corresponding bins. Is this possible? I am not sure if I am missing something.

**Comment by trivialfis** [comment]
I'm not entirely sure what you are trying to suggest but I don't think it's possible to generate bins on the fly without significant performance loss and an overhaul of the library.  We are still trying to optimize memory usage by other means, stay tuned. ;-)

**Comment by deadsoul44** [comment]
I think no memory optimization can be as effective as batch training.

**Comment by trivialfis** [comment]
Batch training implies either the out of core training where the algorithm loops over batches for every tree node, or a stochastic algorithm.  For the former xgboost supports in lastest branch with approx but it's really slow due to IO.  For the latter, GBT is not a stochastic algorithm, you might want to consider neural network instead.

**Comment by deadsoul44** [comment]
tf.data.DataSet has a nice feature of prefetching batches during training.

**Comment by trivialfis** [comment]
Yeah, xgboost's external memory implementation also has support for prefetching, it's just the IO is much more intense than deep learning, as mentioned previously GBT is not a stochastic algorithm and can not use small batches.  My SSD can keep the file reading at about 500MB/s throughout the training, but it's still quite slow compared to in-core training.

**Comment by deadsoul44** [comment]
We can always choose a very large batch size. 
The benefits can justify the penalties at least in some cases. 
For example, using lossguide without depth limit or very large max_depth can achieve the same accuracy with less time. 
Another use case is the problems where many engineered features are required to be added. For example, in time series problems, we can add lagged features more than once. This has the potential to make ARIMA models obsolete.
We might be missing a huge opportunity without batch training.

**Comment by trivialfis** [comment]
@deadsoul44 It's not related to batch size, and lossguide is not recommended for external memory.  Consider the following, if you have 10GB data in binary form, for each tree node, using lossguide, we will loop through the entire dataset.  With a hard drive that can read at 500MB/s, then each tree node will take about 80 seconds as we need to loop through the dataset about 3-4 times depending on the algorithm used.  For a depth 8 tree, you can have maximum of 256 nodes, then it would be 5.7 hours for 1 tree!  `depthwise` is much better as we can build all nodes in one layer at once.  You can try it using the latest master branch, there's a demo in `demo/guide-python/external_memory.py`.

**Comment by PhillipMaire** [comment]
@trivialfis are there speed tests using your SSD you mention 500MB/s but I have no way of understanding how this compares to in core training. I understand that this will depend on many factors like you mentioned above but can you give some idea? are we looking at 2X 10X 100X the time? does grow linearly with increased data size? 

a high quality 1TB SSD is soooo much cheaper than building a PC with 512gb of memory :sweat_smile:
